doc/baremetal/hostlink.rst

```diff
@@ -34,10 +38,55 @@ required. nSIM intercepts those exceptions and handles them. The advantage of
 this approach is that same application binary can be used with other execution
 environments, which also handle software exceptions - unlike the case where a
 system function implementation is really baked inside the application binary.
+To use nSIM GNU IO hostlink in application, add option ``--specs=nsim.specs`` to
+gcc options when linking - library libnsim.a will be linked in and will provide
+implementations to those system level functions. For example, consider simple
+program ``hello.c``:
+
+.. code-block:: c
+
+   #include <stdio.h>
+   int main()
+   {
+       printf("Hello World!\n");
+       return 0;
+   }
+
+Let's build it for ARC HS with nSIM GNU IO Hostlink support and run in nSIM:
+
+.. code-block:: text
+
+   $ arc-elf-gcc -mcpu=hs --specs=nsim.specs ./hello.c -o hello
+   $ nsimdrv -prop=nsim_isa_family=av2hs -prop=nsim_emt=1 ./hello
+   Hello World!
 
-To use hostlink in application, add option ``--specs=nsim.specs`` to gcc options
-when linking - library libnsim.a will be linked in and will provide
-implementations to those system level functions
+Please note ``-prop=nsim_emt=1`` (emulate traps) option which enables nSIM GNU IO
+Hostlink in nSIM. More details can be found in nSIM documentation.
+
+Metaware Hostlink works via memory mailbox named ``__HOSTLINK__``. The running
+application fills this mailbox and then debugger or simulator execute requested
+system call. The advantage of this approach is that binary can be executed in the
+real hardware under Metaware Debugger. The disadvantage is that binary should not
+be stripped to allow debugger to find needed symbols automatically. For
+stripped case it is still possible to pass address manually to debugger or
+simulator, please look into the apropriate documentation for more information.
+To use Metaware Hostlink, add option ``--specs=hl.specs``. Let’s build and run
+our ``hello.c`` with Metaware Hostlink:
+
+.. code-block:: text
+
+   $ arc-elf-gcc -mcpu=hs --specs=hl.specs ./hello.c -o hello
+   $ nsimdrv -prop=nsim_isa_family=av2hs -prop=nsim_hlink_gnu_io_ext=1 ./hello
+   Hello World!
+
+Please note ``-prop=nsim_hlink_gnu_io_ext=1`` option which enables some
+additional system calls support in nSIM. Also make sure that ``-prop=nsim_emt``
+option is disabled in nSIM, because only one Hostlink type can be used at the
```

Comment:
That's a subtle topic, indeed. nSIM documentation lists 3 possible values of `nsim_emt`:
* 0 - disabled, equivalent of not specifying this option on the command line at all
* 1 - enables "GNU hostlink" implemented via exceptions
* 1 - enables "GNU Linux hostlink" which is execution of the application built for ARC Linux on top of nSIM (w/o instantiation of the Linux kernel for ARC)

But I'm not sure if we want to confuse our users that much ;)

Reply:
@abrodkin I originally asked him to add that in hostlink documentation in newlib, which this text has originated from. The reason I asked him to add that was I fell into that trap.